### PR TITLE
WIP: MAISTRA-2426 move FederationStatus into MeshFederation

### PR DESCRIPTION
**Pull request by rcernich**
Signed-off-by: rcernich <rcernich@redhat.com>

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pkg/servicemesh/federation/common/options.go [hunk 1]
```diff
@@ -15,13 +15,32 @@
...
 	ResyncPeriod time.Duration
 	Namespace    string
 }
+
+func (opt ControllerOptions) validate() error {
+	var allErrors []error
+	if opt.KubeClient == nil {
+		allErrors = append(allErrors, fmt.Errorf("the KubeClient field must not be nil"))
+	}
+	if opt.MiastraCS == nil {
```
Comment (shamsher31) [inline pkg/servicemesh/federation/common/options.go]:
```suggestion
	if opt.MaistraCS == nil {
```

File: pkg/servicemesh/federation/common/options.go [hunk 2]
```diff
@@ -15,13 +15,32 @@
...
+		allErrors = append(allErrors, fmt.Errorf("the MiastraCS field must not be nil"))
```
Comment (shamsher31) [inline pkg/servicemesh/federation/common/options.go]:
```suggestion
		allErrors = append(allErrors, fmt.Errorf("the MaistraCS field must not be nil"))
```

File: pkg/servicemesh/federation/common/options.go [hunk 3]
```diff
@@ -15,13 +15,32 @@
...
 	"time"
 
 	"istio.io/istio/pkg/kube"
+	"k8s.io/apimachinery/pkg/util/errors"
+	maistraclient "maistra.io/api/client/versioned"
 )
 
 type ControllerOptions struct {
 	KubeClient   kube.Client
+	MiastraCS    maistraclient.Interface
```
Comment (shamsher31) [inline pkg/servicemesh/federation/common/options.go]:
```suggestion
	MaistraCS    maistraclient.Interface
```

File: pkg/servicemesh/federation/common/resources.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+func NewResourceManager(opts ControllerOptions, mrc memberroll.MemberRollController) (ResourceManager, error) {
+	if err := opts.validate(); err != nil {
+		return nil, err
+	}
+
+	var informerFactory maistraxnsinformer.SharedInformerFactory
+	// Currently, we only watch istio system namespace for MeshFederation resources, which is why this block is disabled.
+	if mrc != nil && false {
+		informerFactory = maistraxnsinformer.NewSharedInformerFactoryWithOptions(opts.MiastraCS, opts.ResyncPeriod, maistraxnsinformer.WithNamespaces())
```
Comment (shamsher31) [inline pkg/servicemesh/federation/common/resources.go]:
```suggestion
		informerFactory = maistraxnsinformer.NewSharedInformerFactoryWithOptions(opts.MaistraCS, opts.ResyncPeriod, maistraxnsinformer.WithNamespaces())
```

File: pkg/servicemesh/federation/common/resources.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		mrc.Register(informerFactory, "federation")
+	} else {
+		informerFactory = maistraxnsinformer.NewSharedInformerFactoryWithOptions(opts.MiastraCS, opts.ResyncPeriod, maistraxnsinformer.WithNamespaces(opts.Namespace))
```
Comment (shamsher31) [inline pkg/servicemesh/federation/common/resources.go]:
```suggestion
		informerFactory = maistraxnsinformer.NewSharedInformerFactoryWithOptions(opts.MaistraCS, opts.ResyncPeriod, maistraxnsinformer.WithNamespaces(opts.Namespace))
```

File: pkg/servicemesh/federation/common/resources.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	}
+	rm := &resourceManager{
+		mcs:  opts.MiastraCS,
```
Comment (shamsher31) [inline pkg/servicemesh/federation/common/resources.go]:
```suggestion
		mcs:  opts.MaistraCS,
```

File: pkg/servicemesh/federation/discovery/controller_test.go [hunk 7]
```diff
@@ -173,40 +207,49 @@ func newTestOptions(discoveryAddress string) options {
...
 	options := newTestOptions("test.address")
-	controller := internalNewController(fake.NewSimpleClientset(), nil, Options{
-		ControllerOptions: common.ControllerOptions{
-			KubeClient:   kube.NewFakeClient(),
-			ResyncPeriod: resyncPeriod,
-			Namespace:    "",
-		},
+	rm, err := common.NewResourceManager(common.ControllerOptions{
+		KubeClient:   kube.NewFakeClient(),
+		MiastraCS:    fake.NewSimpleClientset(),
```
Comment (shamsher31) [inline pkg/servicemesh/federation/discovery/controller_test.go]:
```suggestion
		MaistraCS:    fake.NewSimpleClientset(),
```

File: pkg/servicemesh/federation/federation.go [hunk 8]
```diff
@@ -70,17 +76,30 @@ type Federation struct {
...
-		return nil, err
+		return nil, fmt.Errorf("error creating ClientSet for ServiceMesh: %v", err)
 	}
+	return internalNew(opt, cs)
+}
+
+func internalNew(opt Options, cs maistraclient.Interface) (*Federation, error) {
+	resourceManager, err := common.NewResourceManager(common.ControllerOptions{
+		KubeClient:   opt.KubeClient,
+		MiastraCS:    cs,
```
Comment (shamsher31) [inline pkg/servicemesh/federation/federation.go]:
```suggestion
		MaistraCS:    cs,
```

**Comment by rcernich** [comment]
replaced by #393

**Pull request closed without merging**